feflow/protocols/nonequilibrium_cycling.py

```diff
@@ -202,71 +205,127 @@ def _execute(self, ctx, *, state_a, state_b, mapping, settings, **inputs):
+        state_b_topology, state_b_alchem_resids = _rfe_utils.topologyhelpers.combined_topology(
+            state_a_topology,
+            ligand_b.to_openff().to_topology().to_openmm(),
+            exclude_resids=comp_resids[ligand_a],
         )
 
-        topology_proposals = fe_setup.topology_proposals
-        old_positions = fe_setup.old_positions
-        new_positions = fe_setup.new_positions
-
-        # Generate Hybrid Topology Factory - Generic HTF
-        htf = HybridTopologyFactory(
-            topology_proposal=topology_proposals[phase],
-            current_positions=old_positions[phase],
-            new_positions=new_positions[phase],
-            softcore_LJ_v2=settings.softcore_LJ_v2,
-            interpolate_old_and_new_14s=settings.interpolate_old_and_new_14s,
+        # b. get a list of small molecules for stateB
+        off_mols_state_b = [ligand_b.to_openff(), ]
+        for comp in small_mols_a:
+            if comp != ligand_a:
+                off_mols_state_b.append(comp.to_openff())
+
+        state_b_system = system_generator.create_system(
+            state_b_topology,
+            molecules=off_mols_state_b,
+        )
+
+        #  c. Define correspondence mappings between the two systems
+        ligand_mappings = _rfe_utils.topologyhelpers.get_system_mappings(
+            mapping["ligand"].componentA_to_componentB,
+            state_a_system, state_a_topology, comp_resids[ligand_a],
+            state_b_system, state_b_topology, state_b_alchem_resids,
+            # These are non-optional settings for this method
+            fix_constraints=True,
+        )
+
+        #  d. Finally get the positions
+        state_b_positions = _rfe_utils.topologyhelpers.set_and_check_new_positions(
+            ligand_mappings, state_a_topology, state_b_topology,
+            old_positions=ensure_quantity(state_a_positions, 'openmm'),
+            insert_positions=ensure_quantity(ligand_b.to_openff().conformers[0], 'openmm'),
+        )
+
+        # Get alchemical settings
+        alchemical_settings = settings.alchemical_settings
+
+        # Now we can create the HTF from the previous objects
+        hybrid_factory = HybridTopologyFactory(
+            state_a_system, state_a_positions, state_a_topology,
+            state_b_system, state_b_positions, state_b_topology,
+            old_to_new_atom_map=ligand_mappings['old_to_new_atom_map'],
+            old_to_new_core_atom_map=ligand_mappings['old_to_new_core_atom_map'],
+            use_dispersion_correction=alchemical_settings.use_dispersion_correction,
+            softcore_alpha=alchemical_settings.softcore_alpha,
+            softcore_LJ_v2=alchemical_settings.softcore_LJ_v2,
+            softcore_LJ_v2_alpha=alchemical_settings.softcore_alpha,
+            interpolate_old_and_new_14s=alchemical_settings.interpolate_old_and_new_14s,
+            flatten_torsions=alchemical_settings.flatten_torsions,
         )
+        ####### END OF SETUP #########
```

Comment:
Basically up to here can be moved to a separate unit for #3 and also to help thinking about the openmmtools API points that should be doing these things.

Reply:
Awesome! Thank you for this!